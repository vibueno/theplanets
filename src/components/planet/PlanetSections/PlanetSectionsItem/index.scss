@use 'SRC/variables.scss' as *;
@use 'SRC/mixins.scss' as mixins;

.btn
{
  font: 700 9px/22px Spartan, Arial, Helvetica;

  display: flex;

  margin: 0;

  cursor: pointer;
  text-align: left;
  letter-spacing: 1.93px;
  text-transform: uppercase;

  border: none;
  background-color: $color-blue-dark;

  &[data-selected='false']
  {
    .btn-numbering
    {
      color: $color-grey-light;
    }
  }

  .btn-numbering
  {
    display: none;

    padding: 0 25px 0 0;
  }

  &[data-selected='false']
  {
    .btn-caption
    {
      color: $color-white;
    }
  }

  .btn-caption
  {
    display: inline-block;
  }

  .btn-caption-long
  {
    display: none;

    color: $color-white;
  }
}

.planet-sections-menu
{
  @include mixins.planet-sections-menu-selected;
}

@media only screen and (min-width: 768px)
{
  .planet-sections-btn-panel
  {
    .btn
    {
      .btn-caption
      {
        display: none;
      }

      .btn-caption-long
      {
        display: inline-block;
      }
    }
  }
}
