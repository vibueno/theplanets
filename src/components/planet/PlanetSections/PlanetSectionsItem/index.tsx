import React, { MouseEventHandler } from 'react';

type PlanetSectionsMenuItemProps = {
  sectionKey: string;
  currentSectionKey: string;
  className: string;
  caption: string;
  captionLong: string;
  classNameNumbering: string;
  numbering: string;
  clickHandler: (
    e: React.MouseEvent<HTMLButtonElement>,
    ref: React.RefObject<HTMLButtonElement>
  ) => void;
};

const PlanetSectionsMenuItem = ({
  sectionKey,
  currentSectionKey,
  className,
  caption,
  captionLong,
  classNameNumbering,
  numbering,
  clickHandler
}: PlanetSectionsMenuItemProps) => {
  const ref = React.useRef<HTMLButtonElement>(null);

  return (
    <button
      id={sectionKey}
      className={`btn-section ${className}`}
      onClick={e => clickHandler(e, ref)}
      data-selected={sectionKey === currentSectionKey}
      ref={ref}
    >
      <span className={classNameNumbering}>{numbering}</span>
      <span className="btn-caption">{caption}</span>
      <span className="btn-caption-long">{captionLong}</span>
    </button>
  );
};
export default PlanetSectionsMenuItem;
