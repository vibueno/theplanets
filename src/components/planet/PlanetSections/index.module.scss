@use 'SRC/variables' as *;
@use 'SRC/mixins.scss' as mixins;

@media only screen and (max-width: 767px)
{
  .planetSectionsBtnPanel
  {
    display: none;
  }

  .planetSectionsMenu
  {
    display: grid;

    transition: opacity 1s ease-in-out .5s;

    border-bottom: $border;

    @include mixins.planet-sections-menu-selected;
    grid-row: 2 / 3;
    grid-template-columns: repeat(3, auto);
    &[data-menu-open='false'][data-is-transition-done='true']
    {
      opacity: 1;
    }

    &[data-menu-open='true']
    {
      transition: opacity .25s ease-in-out;

      opacity: 0;
    }
  }
}

@media only screen and (min-width: 767px)
{
  .planetSectionsBtnPanel
  {
    font: 700 12px/25px Spartan, Arial, Helvetica;

    display: flex;
    flex-direction: column;
    justify-content: center;

    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0 0 0 3rem;

    border: none;
  }

  .planetSectionsMenu
  {
    display: none;
  }
}

@media only screen and (min-width: 1200px)
{
  .planetSectionsBtnPanel
  {
    padding: 0;
  }
}

@media only screen and (min-width: 1200px) and (max-height:767px)
{
  .planetSectionsBtnPanel
  {
    margin: 1rem 0 0 0;
  }
}

@media only screen and (min-width: 1200px) and (min-height:768px)
{
  .planetSectionsBtnPanel
  {
    margin: 2rem 0 0 0;
  }
}
