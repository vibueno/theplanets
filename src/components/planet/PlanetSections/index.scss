@use 'SRC/variables' as *;
@use 'SRC/mixins.scss' as mixins;

.btn-section
{
  font: 700 9px/22px Spartan, Arial, Helvetica;

  display: flex;
  align-items: center;

  padding: 5px 25px;

  cursor: pointer;
  letter-spacing: 1.93px;
  text-transform: uppercase;

  background-color: $color-blue-dark;

  .btn-caption
  {
    display: inline-block;
  }

  &[data-selected='false']
  {
    .btn-caption
    {
      color: $color-white;
    }
  }

  .btn-caption-long
  {
    display: none;

    color: $color-white;
  }
}

.planet-sections-menu
{
  display: grid;

  border-bottom: $border;

  @include mixins.planet-sections-menu-selected;
  grid-row: 2 / 3;
  grid-template-columns: repeat(3, auto);

  .btn-section
  {
    position: relative;

    justify-content: center;

    padding: .9rem 0 .75rem 0;

    border: none;

    &::after
    {
      position: absolute;
      bottom: 0;
      left: 0;

      display: block;

      width: 100%;
      height: 5px;

      content: '';
      transition: transform .3s ease;
      transform: scaleX(0);
    }
  }

  #overview
  {
    grid-column: 1 / 2;
  }

  #structure
  {
    grid-column: 2 / 3;
  }

  #geology
  {
    grid-column: 3 / 4;
  }
}

@media only screen and (max-width: 767px)
{
  .btn-numbering
  {
    display: none;
  }
}

@media only screen and (min-width: 768px)
{
  .btn-section
  {
    font: 700 12px/22px Spartan, Arial, Helvetica;

    height: 48px;
    margin: 0 0 1em 0;
    .btn-numbering
    {
      display: inline-block;

      padding: 0 25px 0 0;
    }

    .btn-caption
    {
      display: none;
    }

    .btn-caption-long
    {
      display: inline-block;
    }

    &[data-selected='false']
    {
      transition: background-color .3s;

      border: $border;

      &:hover
      {
        background-color: $color-grey-dark;
      }

      .btn-numbering
      {
        color: $color-grey-light;
      }
    }
  }
}
