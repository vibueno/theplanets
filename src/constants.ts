const sections = {
  overview: {
    id: 'overview',
    number: '01',
    btnCaption: 'overview',
    btnCaptionExtended: 'overview'
  },
  structure: {
    id: 'structure',
    number: '02',
    btnCaption: 'structure',
    btnCaptionExtended: 'internal structure'
  },
  geology: {
    id: 'geology',
    number: '03',
    btnCaption: 'geology',
    btnCaptionExtended: 'surface geology'
  }
};

const planetNames = {
  mercury: 'mercury',
  venus: 'venus',
  earth: 'earth',
  mars: 'mars',
  jupiter: 'jupiter',
  saturn: 'saturn',
  uranus: 'uranus',
  neptune: 'neptune'
};

const planetKeyData = {
  rotation: {
    id: 'rotation',
    title: 'rotation time'
  },
  revolution: {
    id: 'revolution',
    title: 'rotation time'
  },
  radius: {
    id: 'radius',
    title: 'radius'
  },
  temperature: {
    id: 'temperature',
    title: 'average temp.'
  }
};

export { sections, planetNames, planetKeyData };
