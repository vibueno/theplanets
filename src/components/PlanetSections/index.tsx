import React, { useEffect, useRef, MouseEventHandler } from 'react';
import Button from '../../components/Button';

import { sections, sectionKeys } from '../../constants';

type PlanetSectionsProps = {
  planetName: string;
  clickHandler: MouseEventHandler<HTMLButtonElement>;
  cssClass: string;
};

const PlanetSections = ({
  planetName,
  clickHandler,
  cssClass
}: PlanetSectionsProps) => {
  const btnOverviewRef = React.useRef<HTMLButtonElement>(null);

  useEffect(() => {
    btnOverviewRef.current ? btnOverviewRef.current.focus() : null;
  }, [planetName]);

  return (
    <div className={cssClass}>
      {sectionKeys.map(sectionKey => (
        <Button
          id={sections[sectionKey].id}
          cssClass={`btn btn-${planetName}`}
          caption={sections[sectionKey].btnCaption}
          extendedCaption={sections[sectionKey].btnCaptionExtended}
          numberingCssClass="btn-numbering"
          numbering={sections[sectionKey].number}
          clickHandler={clickHandler}
        />
      ))}
    </div>
  );
};

export default PlanetSections;
