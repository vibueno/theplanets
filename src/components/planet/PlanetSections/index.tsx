import React from 'react';
import { useSelector } from 'react-redux';
import { selectors } from 'REDUX/appStateSlice';

import PlanetSectionsItem from './PlanetSectionsItem';

import { PLANETS, SECTIONS, SECTION_KEYS } from 'SRC/constants';

import styles from './index.module.scss';

type PlanetSectionsProps = {
  currentPlanetKey: string;
  currentSectionKey: string;
  clickHandler: (
    e: React.MouseEvent<HTMLButtonElement>,
    ref: React.RefObject<HTMLButtonElement>
  ) => void;
  isMenu: boolean;
};

const PlanetSections = ({
  currentPlanetKey,
  currentSectionKey,
  clickHandler,
  isMenu
}: PlanetSectionsProps) => {
  const isMenuOpen = useSelector(selectors.getIsMenuOpen);
  const isTransitionDone = useSelector(selectors.getIsTransitionDone);

  return (
    <div
      className={`${
        isMenu ? styles.planetSectionsMenu : styles.planetSectionsBtnPanel
      }`}
      data-menu-open={isMenuOpen}
      data-is-transition-done={isTransitionDone}
    >
      {SECTION_KEYS.map(sectionKey => (
        <PlanetSectionsItem
          key={sectionKey}
          sectionKey={sectionKey}
          currentSectionKey={currentSectionKey}
          className={`planetSectionsBtn-${PLANETS[currentPlanetKey].NAME}`}
          caption={SECTIONS[sectionKey].BTN_CAPTION}
          captionLong={SECTIONS[sectionKey].BTN_CAPTION_LONG}
          numbering={SECTIONS[sectionKey].NUMBER}
          clickHandler={clickHandler}
        />
      ))}
    </div>
  );
};

export default PlanetSections;
