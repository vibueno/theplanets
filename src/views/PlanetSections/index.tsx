import React, { useEffect, useRef, MouseEventHandler } from 'react';
import Button from 'COMPONENTS/Button';

import { SECTIONS, SECTION_KEYS } from 'SRC/constants';

type PlanetSectionsProps = {
  planetName: string;
  clickHandler: (
    e: React.MouseEvent<HTMLButtonElement>,
    ref: React.RefObject<HTMLButtonElement>
  ) => void;
  className: string;
};

const PlanetSections = ({
  planetName,
  clickHandler,
  className
}: PlanetSectionsProps) => {
  return (
    <div className={className}>
      {SECTION_KEYS.map(sectionKey => (
        <Button
          key={sectionKey}
          id={SECTIONS[sectionKey].NAME}
          cssClass={`btn btn-${planetName}`}
          caption={SECTIONS[sectionKey].BTN_CAPTION}
          extendedCaption={SECTIONS[sectionKey].BTN_CAPTION_LONG}
          numberingCssClass="btn-numbering"
          numbering={SECTIONS[sectionKey].NUMBER}
          clickHandler={clickHandler}
        />
      ))}
    </div>
  );
};

export default PlanetSections;
