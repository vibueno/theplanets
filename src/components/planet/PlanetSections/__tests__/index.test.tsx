import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import * as renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import PlanetSections from '../';
import { PLANETS, SECTIONS } from 'SRC/constants';

describe('PlanetSections', () => {
  const initialState = {
    appState: {
      isMenuOpen: false,
      isTransitionDone: true,
      currentPlanetKey: PLANETS.MERCURY.KEY,
      currentSectionKey: SECTIONS.OVERVIEW.KEY
    }
  };

  const mockStore = configureStore();

  let store = mockStore(initialState);
  const props = {
    currentPlanetKey: PLANETS.MERCURY.KEY,
    currentSectionKey: SECTIONS.OVERVIEW.KEY,
    clickHandler: () => {},
    isMenu: true
  };

  it('PlanetSections renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <PlanetSections {...props} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('PlanetSections contains as many sections as SECTIONS.SECTIONS', () => {
    render(
      <Provider store={store}>
        <PlanetSections {...props} />
      </Provider>
    );

    expect(screen.queryAllByRole('button').length).toEqual(
      Object.keys(SECTIONS).length
    );
  });
});
