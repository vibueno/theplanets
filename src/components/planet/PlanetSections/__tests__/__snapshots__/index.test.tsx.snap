// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetSections PlanetSections renders correctly 1`] = `
<div
  className="undefined"
  data-is-transition-done={true}
  data-menu-open={false}
>
  <button
    className="undefined undefined"
    data-selected={true}
    id="OVERVIEW"
    onClick={[Function]}
  >
    <span>
      01
    </span>
    <span>
      overview
    </span>
    <span>
      overview
    </span>
  </button>
  <button
    className="undefined undefined"
    data-selected={false}
    id="STRUCTURE"
    onClick={[Function]}
  >
    <span>
      02
    </span>
    <span>
      structure
    </span>
    <span>
      internal structure
    </span>
  </button>
  <button
    className="undefined undefined"
    data-selected={false}
    id="GEOLOGY"
    onClick={[Function]}
  >
    <span>
      03
    </span>
    <span>
      geology
    </span>
    <span>
      surface geology
    </span>
  </button>
</div>
`;
